flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlListState.java

```diff
@@ -103,6 +103,11 @@ public void add(T value) throws Exception {
     @Nullable
     @Override
     public List<TtlValue<T>> getUnexpiredOrNull(@Nonnull List<TtlValue<T>> ttlValues) {
+        // the update operation will clear the whole state if the list becomes empty after init
+        if (ttlValues.isEmpty()) {
+            return ttlValues;
```

Comment:
Can this ever happen here? Are we wrapping the `UserFacingListState` which maps `null` to an empty list, or are empty lists never possible?

Reply:
This could happen before this change(after the elements are all expired). So... if a job restores from a previous Flink version without this change, it may still happen.